// nmtg-template-mailerlite-typography/dist/index.es.js
class Errors {
  ERROR_MESSAGE = (message) => {
    return `Error in component: "${this.component}",  error message: "${message}"`;
  };
  component = '';

  constructor(component = 'No component') {
    this.component = component;
  }

  add(error) {
    const textError = this.ERROR_MESSAGE(error);
    throw new Error(textError);
  }
}

const htmlEscape = string => string
	.replace(/&/g, '&amp;')
	.replace(/"/g, '&quot;')
	.replace(/'/g, '&#39;')
	.replace(/</g, '&lt;')
	.replace(/>/g, '&gt;');

var htmlEscape_1 = (strings, ...values) => {
	if (typeof strings === 'string') {
		return htmlEscape(strings);
	}

	let output = strings[0];
	for (const [index, value] of values.entries()) {
		output = output + htmlEscape(String(value)) + strings[index + 1];
	}

	return output;
};

function stringifyAttributes$1(attributes) {
	const handledAttributes = [];

	for (let [key, value] of Object.entries(attributes)) {
		if (value === false) {
			continue;
		}

		if (Array.isArray(value)) {
			value = value.join(' ');
		}

		let attribute = htmlEscape_1(key);

		if (value !== true) {
			attribute += `="${htmlEscape_1(String(value))}"`;
		}

		handledAttributes.push(attribute);
	}

	return handledAttributes.length > 0 ? ' ' + handledAttributes.join(' ') : '';
}

// import customErrors from './Errors';
// const customError = new customErrors('link');

// import ow from 'ow';

const linkComponent$1 = (params) => {
  // if (typeof params != 'object') {
  //   customError.add('"params" is not "object"');
  // }

  // if (typeof params.attributes == '') {
  //   customError.add('empty attributes');
  // }

  // if (typeof params.content == '') {
  //   customError.add('empty content');
  // }


  const attributesStr = stringifyAttributes$1(attributes);

  const { attributes, content } = params;

  return `<a ${attributesStr}>${content}</a>`;
};

// import Errors from './Errors';

// stringifyAttributes({
//   rainbow: true,
//   number: 1,
//   multiple: ['a', 'b'],
// });


const buttonComponent = (params) => {
  const attributes = {
    id: params.id,
    class: `mlContentButton`,
    href: params.href,
    'data-link-id': params.id,
    style: `font-family: "Poppins", sans-serif; background-color: #d6685e; border-radius: 3px; color: #ffffff; display: inline-block; font-size: 17px; font-weight: 400; line-height: 23px; padding: 15px 0 15px 0; text-align: center; text-decoration: none; width: 260px;`,
  };

  const attributesStr = stringifyAttributes$1(attributes);

  // console.log(attributesStr);

  return linkComponent$1({ content: params.text, attributesStr });
};


// const error = new Errors('strong');

// if (content == '') {
// error.add('No content');
// }

const config = {
  id: '12',
  href: 'google.com'
};
console.log(buttonComponent(config));

// Content Button

// we will probably loose id param during changes that Arthur is doing.
// const buttonMainBlock = (id, href, text) => {

  


//   return `<a class="mlContentButton" href="${href}" data-link-id="${id}" 
//   style="font-family: "Poppins", sans-serif; background-color: #d6685e; border-radius: 3px; color: #ffffff; display: inline-block; font-size: 17px; font-weight: 400; line-height: 23px; padding: 15px 0 15px 0; text-align: center; text-decoration: none; width: 260px;" target="_self">
//   ${text}
//   </a>`;
// };

const buttonComponentOld = (params) => {
  const { id, href, text } = params;
  const error = new Errors('contentButton');

  if (id == '') {
    error.add('No id button');
  }
  if (href == '') {
    error.add('No href button');
  }
  if (text == '') {
    error.add('No text button');
  }

  const attributes = {
    class: `mlContentButton`,
    href: href,
    'data-link-id': id,
    style: `font-family: "Poppins", sans-serif; background-color: #d6685e; border-radius: 3px; color: #ffffff; display: inline-block; font-size: 17px; font-weight: 400; line-height: 23px; padding: 15px 0 15px 0; text-align: center; text-decoration: none; width: 260px;`,
  };

  const html = buttonComponent({ content: text, attributes });
  console.log(html);
  return html;

//   // const attributesStr = stringifyAttributes(attributes);
//   // console.log(attributesStr);
//   // return linkComponent({ content: params.text, attributes });
  // return buttonMainBlock(id, href, text);

};

// Component heading

// import Errors from './Errors';

// TODO replace with a link component
const headingMainBlock = (hrefTitle, idTitle, textTitle) => {

  // const aHEADNMTG = headingNMTG({
  //   href: hrefTitle,
  //   id: idTitle
  // })
  // console.log(aHEADNMTG);

  const attributes = {
    target: `_self`,
    href: params.hrefTitle,
    'data-link-id': params.idTitle,
    style: `text-decoration: none; color: #111111; text-transform: none; font-style: normal; text-decoration: none;`,
  };

  const attributesStr = stringifyAttributes(attributes);

  // console.log(attributesStr);

  return linkComponent$1({ content: textTitle, attributesStr });
};

// export default function (params) {
//   const { hrefTitle, idTitle, textTitle } = params;
//   const error = new Errors('contentTitleText');

//   if (hrefTitle == '') {
//     error.add('No hrefTitle');
//   }
//   if (idTitle == '') {
//     error.add('No idTitle');
//   }
//   if (textTitle == '') {
//     error.add('No textTitle');
//   }

//   return headingMainBlock(hrefTitle, idTitle, textTitle);
// }

// Create strong component

const strongMainBlock = (content) => {
  return `<strong>${content}</strong>`;
};

function strong (content) {
  const error = new Errors('strong');

  if (content == '') {
    error.add('No content');
  }

  return strongMainBlock(content);
}

// Component contentSubTitleText

const subHeadingMainBlock = (subTitleText) => {
  return `<td align="center" class="bodyTitle" style="font-family: 'Poppins', sans-serif; font-size: 19px; font-weight: 400; line-height: 150%; color: #111111; text-transform: none; font-style: normal; text-decoration: none; text-align: center;">${subTitleText}</td>`;
};

const renderSubHeading = (params) => {
  const { subTitleText } = params;
  const error = new Errors('contentSubTitleText');

  if (subTitleText == '') {
    error.add('No subTitleText');
  }

  return subHeadingMainBlock(subTitleText);
};

// <span style="color: #111111;">Unsubscribe</span>
const spanHeadingBlock = (title) => {
  return `<span style="color: #111111;">${title}</span>`;
};

const customError$1 = new Errors('paragraph');

const paragraphComponent = (params) => {
  if (typeof params != 'object') {
    customError$1.add('"params" is not "object"');
  }

  if (typeof params.attributes == '') {
    customError$1.add('empty attributes');
  }

  if (typeof params.content == '') {
    customError$1.add('empty content');
  }

  const { attributes, content } = params;

  return `<p ${attributes}>${content}</p>`;
};

const customError = new Errors('link');

const linkComponent = (params) => {
  if (typeof params != 'object') {
    customError.add('"params" is not "object"');
  }

  if (typeof params.attributes == '') {
    customError.add('empty attributes');
  }

  if (typeof params.content == '') {
    customError.add('empty content');
  }

  const { attributes, content } = params;

  return `<a ${attributes}>${content}</a>`;
};

const separatorComponent = () => {
    return `<table align="center" border="0" bgcolor="#ffffff" class="mlContentTable mlContentTableDefault" cellpadding="0" cellspacing="0" width="640">
    <tbody><tr>
      <td class="mlContentTableCardTd">
        <table align="center" bgcolor="#ffffff" border="0" cellpadding="0" cellspacing="0" class="mlContentTable ml-default" style="width: 640px; min-width: 640px;" width="640">
          <tbody><tr>
            <td>
              <table role="presentation" cellpadding="0" cellspacing="0" border="0" align="center" width="640" style="width: 640px; min-width: 640px;" class="mlContentTable">
                <tbody><tr>
                  <td height="20" class="spacingHeight-20" style="line-height: 20px; min-height: 20px;"></td>
                </tr>
              </tbody></table>
              <table role="presentation" cellpadding="0" cellspacing="0" border="0" align="center" width="640" style="width: 640px; min-width: 640px;" class="mlContentTable">
                <tbody><tr>
                  <td align="center" class="">
                    <table role="presentation" cellpadding="0" cellspacing="0" border="0" align="center" width="100%" style="border-top: 1px solid #ededf3; border-collapse: initial;" class="">
                      <tbody><tr>
                        <td height="20" class="spacingHeight-20" style="line-height: 20px; min-height: 20px;"></td>
                      </tr>
                    </tbody></table>
                  </td>
                </tr>
              </tbody></table>
            </td>
          </tr>
        </tbody></table>
      </td>
    </tr>
    </tbody></table>`;
};

// Error

// if (typeof params != 'object') {
//   customError.add('"params" is not "object"');
// }

// if (typeof params.attributes == '') {
//   customError.add('empty attributes');
// }
const imageComponent = (attributes) => {
  // const { src, width } = attributes;

  // if(!width){
  //   return  `<img src="${src}" border="0" alt=""  style="display: block;">`;
  // }


  // return `<img src="${src}" border="0" alt="" width="${width}" style="display: block;">`;

  return `<img ${attributes}>`;
};
// return `<img src="${src}" border="0" alt="" width="${width}" style="display: block;">`;

//function createPathToImage(src) {
//  return pathToImages + src;
//}

const imgWrapper = (params) => {
  const { id, href, src, width } = params;

  const error = new Errors('image');

  if (id == '') {
    error.add('No id');
  }
  if (href == '') {
    error.add('No href');
  }
  if (src == '') {
    error.add('No src');
  }

  return imageMainBlock(id, href, src, width);
};

// Main image block start
// const obj = {
//   src: createPathToImage(src),
//   width: 267, // TODO check if width must be string
// };

// TODO move into Link + Image component
const imageMainBlock = (id, href, src, width) => {


  // const imagePath = ;

  const attr = `src="${src}" 
    border="0" alt="" ` + 
    (width === undefined ? `` : `width="${width}"`) +
     ` style="display: block;"`;
     
  const image = imageComponent(attr);

  const params = {
    attributes: `href="${href}" data-link-id="${id}" target="_self"`,
    content: image,
  };
  return linkComponent(params);
};

buttonComponentOld({
  id: '12',
  href: 'google.com',
  text: 'GoooGLE link'
});
console.log(buttonComponentOld);

export { buttonComponentOld, headingMainBlock as heading, imageComponent, imgWrapper, linkComponent, paragraphComponent, renderSubHeading, separatorComponent, spanHeadingBlock, strong };
