// email-template-object/src/factoryThree.js
// checks should be runned inside display
const displayFactory = (settings, checks = [], isError = false) => {

    // console.log(settings)

    const { component, params, subcomponents } = settings;

    if(!component) throw new Error('no component was passed')

    if(!params) throw new Error('no params was passed')
    
    // if(!subcomponents) throw new Error('no component was passed')
    


    return {
      error: false, 
      params: params,
      subcomponents: (subcomponents) ? subcomponents : null,  
      // log: false,

      display: (params, subcomponents) => { 
  
          //here i want to apply a check and see if everything is fine, if not we generate error = true 
        //   try {
        //     component(params, subcomponents) 
        //   } catch (err) {
        //     // statements to handle any exceptions
        //     console.log(err);
        //     // this.error = true;
        //   }

          
      
      },
        
      checks: () => { return []; },  
        
        
      init: () => {
        let COM;


      },
      
      try: () => {
  
      },

	  get: () => { return component; },

      checks: () => {
          checks.map(() => {

          })

          if(subcomponents){
            //   subcomponents.map((component) => {
            //     component.init();
            //   })
          }

        

      },

      log: () => { 
          console.log('123');

          console.log(this);
          
      // //     console.log(this.display()) 
      },



      // I want to have a better way to report errors, so each callback can report an issue, so we can track it better 
      isError: () => { if(this.error){ return 'errror is here'; } }
    }
  }


  
  export default displayFactoryThree;
